solr/core/src/java/org/apache/solr/servlet/HttpSolrCall.java

```diff
@@ -880,37 +891,43 @@ private SolrCore checkProps(ZkNodeProps zkProps) {
     return core;
   }
 
-  private void getSlicesForCollections(ClusterState clusterState,
+  private int getSlicesAndReplicationFactorForCollections(ClusterState clusterState,
                                        Collection<Slice> slices, boolean activeSlices) {
+    int replicationFactor = 3;
     if (activeSlices) {
       for (Map.Entry<String, DocCollection> entry : clusterState.getCollectionsMap().entrySet()) {
         final Collection<Slice> activeCollectionSlices = entry.getValue().getActiveSlices();
+        replicationFactor = entry.getValue().getReplicationFactor();
         if (activeCollectionSlices != null) {
           slices.addAll(activeCollectionSlices);
         }
       }
     } else {
       for (Map.Entry<String, DocCollection> entry : clusterState.getCollectionsMap().entrySet()) {
         final Collection<Slice> collectionSlices = entry.getValue().getSlices();
+        replicationFactor = entry.getValue().getReplicationFactor();
         if (collectionSlices != null) {
           slices.addAll(collectionSlices);
         }
       }
     }
+    return replicationFactor;
   }
 
   private String getRemotCoreUrl(String collectionName, String origCorename) {
     ClusterState clusterState = cores.getZkController().getClusterState();
     Collection<Slice> slices = clusterState.getActiveSlices(collectionName);
+    final DocCollection docCollection = clusterState.getCollectionOrNull(collectionName);
+    int replicationFactor = (docCollection != null) ? docCollection.getReplicationFactor() : 0;
     boolean byCoreName = false;
 
     if (slices == null) {
       slices = new ArrayList<>();
       // look by core name
       byCoreName = true;
-      getSlicesForCollections(clusterState, slices, true);
+      replicationFactor = getSlicesAndReplicationFactorForCollections(clusterState, slices, true);
       if (slices.isEmpty()) {
-        getSlicesForCollections(clusterState, slices, false);
+        replicationFactor = getSlicesAndReplicationFactorForCollections(clusterState, slices, false);
       }
```

Comment:
Personally, I would rather not have this condition at all than limiting number of forwarded requests.

Reply:
This was done to forward requests more aggressively and avoid missing out slices in case it's recovering or some other reason inconsistent state comes up.
To limit impact I have considered limiting requests to replica count